From 5967798a47ba8835ae100c01a8d647d2f9a2412f Mon Sep 17 00:00:00 2001
From: Rico Tzschichholz <ricotz@ubuntu.com>
Date: Wed, 27 May 2020 17:28:27 +0200
Subject: applications: Properly handle possible error of
 GMenu.Tree.load_sync()

Prevent crash on "g_mutex_clear() called on uninitialised or locked mutex"
---
 docklets/Applications/ApplicationsDockItem.vala | 15 ++++++++++++++-
 1 file changed, 14 insertions(+), 1 deletion(-)

diff --git a/docklets/Applications/ApplicationsDockItem.vala b/docklets/Applications/ApplicationsDockItem.vala
index dcb6cd1..67d17ff 100644
--- a/docklets/Applications/ApplicationsDockItem.vala
+++ b/docklets/Applications/ApplicationsDockItem.vala
@@ -23,6 +23,7 @@ namespace Docky
 	{
 		GMenu.Tree apps_menu;
 		Mutex apps_menu_mutex;
+		bool apps_loaded;
 
 		/**
 		 * {@inheritDoc}
@@ -52,7 +53,13 @@ namespace Docky
 		{
 			Worker.get_default ().add_task_with_result.begin<void*> (() => {
 				apps_menu_mutex.lock ();
-				apps_menu.load_sync ();
+				try {
+					apps_menu.load_sync ();
+					apps_loaded = true;
+				} catch (Error e) {
+					warning ("Failed to load applications (%s)", e.message);
+					apps_loaded = false;
+				}
 				apps_menu_mutex.unlock ();
 				return null;
 			}, TaskPriority.HIGH);
@@ -72,6 +79,12 @@ namespace Docky
 		{
 			var items = new Gee.ArrayList<Gtk.MenuItem> ();
 
+			if (!apps_loaded) {
+				var item = create_menu_item (_("No applications available"), null, false);
+				items.add (item);
+				return items;
+			}
+
 			var iter = apps_menu.get_root_directory ().iter ();
 			GMenu.TreeItemType type;
 			while ((type = iter.next ()) != GMenu.TreeItemType.INVALID) {
-- 
cgit v1.1

